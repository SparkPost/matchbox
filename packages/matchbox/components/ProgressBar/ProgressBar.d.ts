export const StyledProgressBarOuter: any;
export const StyledProgressBarInner: any;
export default ProgressBar;
declare const ProgressBar: any;
