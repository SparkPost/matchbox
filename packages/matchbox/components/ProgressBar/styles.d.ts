export function outerBase(): string;
export function innerBase(): string;
export function visualSize(props: any): "\n        height: 0.25rem;\n      " | "\n        height: 0.75rem;\n      ";
export function calculatedWidth(props: any): string;
