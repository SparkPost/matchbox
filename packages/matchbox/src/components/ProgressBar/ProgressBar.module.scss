@import '../../styles/config';

.ProgressBar {
  position: relative;
  margin-bottom: rem(18);
  height: rem(21);
  background: color(gray, 8);
  border-bottom: 1px solid color(gray, 7);
  border-radius: border-radius();
}

.Progress {
  height: rem(21);
  border-radius: border-radius();
  box-shadow: shadow();

  .orange & {
    background: color(orange);
    border-bottom: 1px solid color(orange, dark);
  }

  .blue & {
    background: color(blue);
    border-bottom: 1px solid color(blue, dark);
  }

  .navy & {
    background: color(navy);
    border-bottom: 1px solid color(navy, dark);
  }
  .purple & {
    background: color(purple);
    border-bottom: 1px solid color(purple, dark);
  }
  .red & {
    background: color(red);
    border-bottom: 1px solid color(red, dark);
  }
}
