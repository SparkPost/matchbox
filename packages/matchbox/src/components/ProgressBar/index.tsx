export { default as ProgressBar } from './ProgressBar';
export * from './ProgressBar';
