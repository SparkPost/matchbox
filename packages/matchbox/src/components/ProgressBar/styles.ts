import { tokens } from '@sparkpost/design-tokens';

export const outerBase = () => `
  position: relative;
  background-color: ${tokens.color_gray_300};
  border-radius: ${tokens.borderRadius_100};
  width: 100%;
`;

export const innerBase = () => `
  background-color: ${tokens.color_blue_700};
  border-radius: ${tokens.borderRadius_100} ${tokens.borderRadius_0} ${tokens.borderRadius_0} ${tokens.borderRadius_100};
`;

export const visualSize = (props) => {
  switch (props.visualSize) {
    case 'small':
      return `
        height: 0.25rem;
      `;
    case 'normal':
    default:
      return `
        height: 0.75rem;
      `;
  }
};

export const calculatedWidth = (props) => {
  let percentage = props.completed;

  if (percentage > 100) {
    percentage = 100;
  } else if (percentage < 1) {
    percentage = 0;
  }

  return `width: ${percentage}%`;
};
