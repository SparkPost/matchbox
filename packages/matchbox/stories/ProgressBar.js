import React from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';
import { action } from '@storybook/addon-actions';
import { StoryContainer } from './helpers';

import { ProgressBar } from '../src';

export default storiesOf('ProgressBar', module)
  .addDecorator((getStory) => (
    <StoryContainer>{ getStory() }</StoryContainer>
  ))
  .add('Default', withInfo()(() => (
    <ProgressBar completed={54} color='red' />
  )));
