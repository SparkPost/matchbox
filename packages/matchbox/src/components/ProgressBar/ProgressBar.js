import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import styles from './ProgressBar.module.scss';

class ProgressBar extends Component {
  static displayName = 'ProgressBar';

  static propTypes = {
    /**
     * Completion in percentage
     */
    completed: PropTypes.number.isRequired,

    /**
     * Bar color
     */
    color: PropTypes.oneOf(['orange', 'blue', 'navy', 'purple', 'red'])
  };

  static defaultProps = {
    completed: 0,
    color: 'orange'
  }

  render() {
    const { completed = 0, color } = this.props;

    let percentage = completed;

    if (percentage > 100) {
      percentage = 100;
    } else if (percentage < 1) {
      percentage = 0;
    }

    return (
      <div className={classnames(styles.ProgressBar, styles[color])}>
        <div className={styles.Progress} style={{ width: `${percentage}%` }}/>
      </div>
    );
  }
}

export default ProgressBar;
