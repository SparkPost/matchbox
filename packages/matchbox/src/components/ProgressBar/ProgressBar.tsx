import React from 'react';
import styled from 'styled-components';
import { margin, MarginProps, compose } from 'styled-system';
import { pick } from '../../helpers/props';
import { Box, BoxProps } from '../Box';

import { outerBase, innerBase, visualSize, calculatedWidth } from './styles';

const system = compose(margin);

export const StyledProgressBarOuter = styled(Box)<BoxProps & { $visualSize?: string }>`
  ${outerBase}
  ${visualSize}
  ${system}
`;

export const StyledProgressBarInner = styled(Box)<
  BoxProps & { $visualSize?: string; $completed?: number }
>`
  ${innerBase}
  ${visualSize}
  ${calculatedWidth}
`;

export type ProgressBarProps = MarginProps & {
  completed: number;

  /**
   * @deprecated Color is always blue for now. This may be updated in the future.
   */
  color?: 'orange' | 'blue' | 'navy' | 'purple' | 'red';
  'data-id'?: string;
  id?: string;
  size?: 'normal' | 'small';
  /**
   * Describes what the progressbar represents - content is visually hidden
   */
  label?: string;

  /**
   * Used to describe the current status of the progress bar to screen reader only users if completion percentage is inadequete
   */
  valueText?: string;
};

const ProgressBar = React.forwardRef<HTMLDivElement, ProgressBarProps>(function ProgressBar(
  props,
  userRef,
) {
  const {
    completed = 0,
    label,
    size = 'normal',
    valueText,
    id,
    'data-id': dataId,
    ...rest
  } = props;
  const systemProps = pick(rest, system.propNames);

  return (
    <StyledProgressBarOuter
      $visualSize={size}
      role="progressbar"
      aria-label={label}
      aria-valuenow={completed}
      aria-valuemin={0}
      aria-valuemax={100}
      aria-valuetext={valueText}
      {...systemProps}
      data-id={dataId}
      id={id}
      ref={userRef}
    >
      <StyledProgressBarInner $completed={completed} $visualSize={size} />
    </StyledProgressBarOuter>
  );
});

ProgressBar.displayName = 'ProgressBar';

export default ProgressBar;
