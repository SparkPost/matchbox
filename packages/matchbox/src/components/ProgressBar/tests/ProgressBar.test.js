import React from 'react';
import { shallow } from 'enzyme';

import ProgressBar from '../ProgressBar';

describe('ProgressBar', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<ProgressBar />);
  });

  it('renders progressbar at defaults', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders at 50% progress in navy color', () => {
    wrapper.setProps({ completed: 50, color: 'navy' });
    expect(wrapper).toMatchSnapshot();
  });

  it('renders at 100% progress if given a value over 100', () => {
    wrapper.setProps({ completed: 200 });
    expect(wrapper.find('.Progress').props().style.width).toEqual('100%');
  });
});
