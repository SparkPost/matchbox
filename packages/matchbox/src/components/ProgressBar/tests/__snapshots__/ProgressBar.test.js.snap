// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar renders at 50% progress in navy color 1`] = `
<div
  className="ProgressBar navy"
>
  <div
    className="Progress"
    style={
      Object {
        "width": "50%",
      }
    }
  />
</div>
`;

exports[`ProgressBar renders progressbar at defaults 1`] = `
<div
  className="ProgressBar orange"
>
  <div
    className="Progress"
    style={
      Object {
        "width": "0%",
      }
    }
  />
</div>
`;
